### PR TITLE
Kevin/vxmark add backend store

### PR DESCRIPTION
**Pull request by kshen0**
## Overview
Can follow commit-by-commit

https://github.com/votingworks/vxsuite/issues/3084

* Adds a store to VxMark
* Factors out some test utils for mock USBs and ballot package read/write from `scan` to `libs`
* Adds get/set methods for system settings to VxMark API

## Demo Video or Screenshot
N/A no user facing change

## Testing Plan
* added automated tests

## Checklist

- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced. <-- I'll do this once all ballot package logic has been centralized
- [x] I have added JSDoc comments to any newly introduced exports

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates

### CODE REVIEW

File: libs/backend/src/ballot_package/ballot_package_io.test.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  );
+  assert(ballotPackageResult.isOk());
+  const ballotPackage = ballotPackageResult.ok();
+  expect(ballotPackage.electionDefinition).toEqual(electionDefinition);
+  expect(ballotPackage.systemSettings).toEqual(
+    safeParseSystemSettings(systemSettings.asText()).unsafeUnwrap()
+  );
+});
+
+test("readBallotPackageFromUsb uses default system settings when system settings don't exist in the zip file", async () => {
```
Comment (kshen0) [inline libs/backend/src/ballot_package/ballot_package_io.test.ts:56]: Some of these tests are ~direct ports from https://github.com/votingworks/vxsuite/pull/3223/commits/ac709203fb5a83b492784ca318b98091d6a940a1#diff-ec10310b6ce5e9cd2721ee574ce861f62cab75215f744ae2b105c716492b3583L130

But there are also a few net new ones. Test descriptions should match up pretty closely

File: libs/backend/src/ballot_package/ballot_package_io.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ * @param usbDrive UsbDrive representing status of an inserted USB drive
+ * @param logger an instance of Logger
+ * @returns Result<BallotPackage, BallotPackageConfigurationError> intended to be consumed by an API handler
+ */
+export async function readBallotPackageFromUsb(
+  authStatus: AuthStatus,
+  usbDrive: UsbDrive,
+  logger: Logger
+): Promise<Result<BallotPackage, BallotPackageConfigurationError>> {
+  // The frontend tries to prevent ballot package configuration attempts until an election
```
Comment (kshen0) [inline libs/backend/src/ballot_package/ballot_package_io.ts:28]: Ported from https://github.com/votingworks/vxsuite/pull/3223/commits/ac709203fb5a83b492784ca318b98091d6a940a1#diff-80d3b0dac90c1c3370e2a5dc4fa423885f5fbd2c9ac5745f34e01057a8820fb2L99

File: libs/backend/src/ballot_package/test_utils.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+import { assert } from '@votingworks/basics';
+import { execSync } from 'child_process';
+import { join } from 'path';
+import * as fsSync from 'fs';
+import {
+  DEFAULT_SYSTEM_SETTINGS,
+  ElectionDefinition,
+} from '@votingworks/types';
+import { Buffer } from 'buffer';
+
```
Comment (kshen0) [inline libs/backend/src/ballot_package/test_utils.ts:12]: Ported from `apps/scan/backend/test/helpers/shared_helpers.ts`

File: apps/mark/backend/src/app.test.ts [hunk 4]
```diff
@@ -24,29 +30,31 @@ test('uses machine config from env', async () => {
...
-  const { apiClient } = createApp();
+  const { apiClient, server } = createApp();
   expect(await apiClient.getMachineConfig()).toEqual({
     machineId: 'test-machine-id',
     codeVersion: 'test-code-version',
     screenOrientation: 'landscape',
   });
 
   process.env = originalEnv;
+  server.close();
```
Comment (kshen0) [inline apps/mark/backend/src/app.test.ts:41]: Not closing the server was preventing Jest from ending gracefully

File: apps/mark/backend/src/app.ts [hunk 5]
```diff
@@ -70,6 +80,36 @@ function buildApi(auth: InsertedSmartCardAuthApi) {
       return auth.endCardlessVoterSession(constructAuthMachineState(input));
     },
 
+    getSystemSettings(): SystemSettings | undefined {
+      return workspace.store.getSystemSettings();
+    },
+
+    async configureBallotPackageFromUsb(input: {
```
Comment (kshen0) [inline apps/mark/backend/src/app.ts:87]: This function is tested but uncalled from production code

File: apps/mark/backend/src/globals.ts [hunk 6]
```diff
@@ -21,3 +22,12 @@ export const NODE_ENV = unsafeParse(
   NodeEnvSchema,
   process.env.NODE_ENV ?? 'development'
 );
+
+/**
+ * Where should the database and audio files go?
```
Comment (kshen0) [inline apps/mark/backend/src/globals.ts:27]: Assuming we are adding audio files later

File: apps/mark/backend/schema.sql [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  is_test_mode boolean not null default true,
+  skip_election_hash_check boolean not null default false,
+  polls_state text not null default "polls_closed_initial",
+  ballot_count_when_ballot_bag_last_replaced integer not null default 0,
+  is_sound_muted boolean not null default false,
+  marginal_mark_threshold_override real,
+  definite_mark_threshold_override real,
+  cvrs_backed_up_at datetime,
+  scanner_backed_up_at datetime,
+  created_at timestamp not null default current_timestamp
```
Comment (arsalansufi) [inline apps/mark/backend/schema.sql:15]: I think there are some extra fields copied from VxScan or VxCentralScan that we don't need, e.g `marginal_mark_threshold_override` and `ballot_count_when_ballot_bag_last_replaced`. Might make sense to trim down to just what we're using right now.

Comment (kshen0) [reply apps/mark/backend/schema.sql:5]: [267511e](https://github.com/votingworks/vxsuite/pull/3223/commits/267511edc5ea8b3e91a9e8fefa26c0bcd2d3ae9a)

**Comment by kshen0** [comment]
@arsalansufi re: `server.close()`. I updated in [4d5ba0b](https://github.com/votingworks/vxsuite/pull/3223/commits/4d5ba0b5a8e2becf2b5195d018f7861f5e90921f) which has the disadvantage that destructuring into a global var requires separating declaration and assignment. So you can either

1. make all variables from `createApp()` global. This means you risk forgetting to destructure eg. `apiMock` in a test, and then you're at risk of using an old `apiMock` from another test.
2. make some vars global and define some in test scope like below, which is about as clunky as calling `server.close()` explicitly.

```
let server: Server;
afterEach(() => { server.close() });
...
test('test something', () => {
  const apiMock;
  ({apiMock, server} = createApp(...));
}
```

If you know of a cleaner way to do this while still taking advantage of `afterEach` let me know!

File: libs/logging/src/log_event_ids.ts [hunk 8]
```diff
@@ -108,7 +108,7 @@ export enum LogEventId {
   PrepareBootFromUsbComplete = 'prepare-boot-from-usb-complete',
   RebootMachine = 'reboot-machine',
   // VxScan/VxCentralScan logs
```
[the remark this answers is not in the document]
Comment (kshen0) [reply libs/logging/src/log_event_ids.ts]: [91b787c](https://github.com/votingworks/vxsuite/pull/3223/commits/91b787ca35705bb263bc8b53e20e549fedf736d6)

File: libs/utils/tsconfig.json [hunk 9]
```diff
@@ -12,7 +12,8 @@
     "noFallthroughCasesInSwitch": true,
     "esModuleInterop": true,
     "skipLibCheck": true,
-    "resolveJsonModule": true
+    "resolveJsonModule": true,
+    "noPropertyAccessFromIndexSignature": false
```
[the remark this answers is not in the document]
Comment (kshen0) [reply libs/utils/tsconfig.json]: I was getting the error `[Property 'foo' comes from an index signature, so it must be accessed with ['foo']]` (can't recall which property).

But I can't repro after removing in [91b787c](https://github.com/votingworks/vxsuite/pull/3223/commits/91b787ca35705bb263bc8b53e20e549fedf736d6), so I'll leave it out

File: libs/backend/src/ballot_package/ballot_package_io.test.ts [hunk 10]
```diff
@@ -25,6 +25,7 @@ test('readBallotPackageFromUsb can read a ballot package from usb', async () =>
     user: fakeElectionManagerUser({
       electionHash: electionDefinition.electionHash,
     }),
+    sessionExpiresAt: new Date().getTime() + 60 * 1000,
```
Comment (arsalansufi) [inline libs/backend/src/ballot_package/ballot_package_io.test.ts:28]: This is of course fine as is, but thought I'd mention that test-utils provides a `fakeSessionExpiresAt()` helper

File: apps/mark/backend/src/app.test.ts [hunk 11]
```diff
@@ -31,31 +47,29 @@ test('uses machine config from env', async () => {
     VX_SCREEN_ORIENTATION: 'landscape',
   };
 
-  const { apiClient, server } = createApp();
+  ({ apiClient, server } = createApp());
```
Comment (arsalansufi) [inline apps/mark/backend/src/app.test.ts:50]: To your comment regarding making it possible to use an `afterEach` for server closing, the main alternative I can think of is making all the variables global but defining them in a `beforeEach`, e.g.

```ts
let apiClient: grout.Client<Api>;
let mockAuth: InsertedSmartCardAuthApi;
let mockUsb: MockUsb;
let server: Server;

beforeEach(() => {
  ({ apiClient, mockAuth, mockUsb, server } = createApp());
});

afterEach(() => {
  server.close();
});
```

This should make this concern you mention impossible:

> This means you risk forgetting to destructure eg. apiMock in a test, and then you're at risk of using an old apiMock from another test.

**Review by arsalansufi — APPROVED** [review]
A few last thoughts then LGTM!

File: apps/mark/backend/src/app.test.ts (hunk 11 above)
Comment (kshen0) [reply apps/mark/backend/src/app.test.ts]: [361e658](https://github.com/votingworks/vxsuite/pull/3223/commits/361e6588a0ef7f36b2b04ecc48d125a8dddb1275)

Great point. Had it in my head that we should only be destructuring exactly the fields we need, which in retrospect is a constraint that gains us nothing.

**Pull request merged**